README.md

```diff
@@ -50,9 +50,200 @@ Then, export the openai env vars `OPENAI_API_BASE` and `OPENAI_API_KEY`. Remembe
+  model: microsoft/phi-2
+  api_key: ${PHI2_API_KEY}
+  api_base: ${PHI2_API_BASE}
+  api_type: null
+  api_version: null
+  max_tokens: 256
+  temperature: 0.0
+
+- name: gpt-bwd
+  model: "gpt-35-turbo-instruct"
+  api_key: ${GPT_API_KEY}
+  api_base: ${GPT_API_BASE}
+  api_type: ${GPT_API_TYPE}
+  api_version: ${GPT_API_VERSION}
+  temperature: 0.7
+  max_tokens: 512
+
+```
+
+Then, you can register the models using the `register_from_yaml` method and get them using the `get` method, as follows:
+
+```python
+llm_registry = LLMRegistry.from_yaml("connections.yaml")
+fwd_model = llm_registry.get("phi2-fwd")
+bwd_model = llm_registry.get("gpt-bwd")
+
+output = bwd_model.generate("Why do programmers prefer dark mode?")
+
+# You can always provide extra keyword arguments to the `generate` method,
+# which will override the ones provided when instantiating the models.
+
+output = bwd_model.generate(
+    "Why do programmers prefer dark mode?",
+    max_tokens=100,
+    echo=True,
+)
+```
+
+## Losses, Samplers and Scores
+
+DLN provides a few losses that can be found in [dln/losses.py](dln/loss.py). A simple example of how to use them is as follows:
+
+```python
+from dln.loss import LossRegistry
+from dln.postprocessing import postprocess_prediction
+
+LossRegistry.available_losses()  # list available losses
+loss_fn = LossRegistry.instantiate(
+    "exact_match_loss", postprocess_prediction
+)
+losses = loss_fn(y_hat, y)
+```
+For sampling and scoring both prompts and hidden states for the Variational Inference algorithm, samplers are found in [dln/vi/sampler.py](dln/vi/sampler.py), and the LogProbsScore in [dln/score.py](dln/score.py). Samplers use templates that are found in [dln/templates.py](dln/templates.py).
+
+
+```python
+import numpy as np
+from dln.operator import LLMRegistry
+from dln.vi.sampler import PosteriorSampler, PromptSampler
+from dln.score import LogProbsScore, ScoreRequest
+
+llm_registry = LLMRegistry()
+llm = llm_registry.register(
+    "llm",
+    "microsoft/phi-2",
+)
+
+prompt_sampler = PromptSampler(llm, "q_action_prompt")
+posterior_sampler = PosteriorSampler(llm, "suffix_forward_tbs")
+logprobs_score = LogProbsScore(llm)
+
+prompt_proposals = prompt_sampler.sample_q_p(
+    inputs=["France", "Canada", "Brazil"],
+    y=["Paris", "Ottawa", "Brasilia"],
+    y_hat=["Paris", "Ottawa", "Sao Paulo"],
+    losses=[0, 0, 0, 1],
+    prompt="What is the capital of this country",
+    num_samples=10,
+)  # sample prompts
+
+hidden_states = posterior_sampler.sample_q_h(
+    x=np.array(["France", "Canada", "Brazil"]),
+    y=["Paris", "Ottawa", "Brasilia"],
+    h=["Paris", "Toronto", "Sao Paulo"],
+    prompt="What is the largest city in this country",
+    next_prompt="What is the capital of this country",
+    num_samples=10,
+)
+
+score_request = ScoreRequest(
+    context="What is the capital of this country: Canada",
+    target="Ottawa",
+    payload="Ottawa",
+)
+score = logprobs_score.score_requests([score_request])
```

Comment:
Got this error when running this but I see you have listed it as a todo.
```
InvalidRequestError: Setting 'echo' and 'logprobs' at the same time is not supported for this model.
```

Reply:
Actually, I will do it in this PR. Somewhere in the readme add a warning mentioning that people can encounter this error if using openai models.

Reply:
Added warning message to main and vi README files